drivers/openstack/client.go

```diff
@@ -48,44 +52,124 @@ type Client interface {
 	GetFloatingIPs(d *Driver) ([]FloatingIP, error)
 	GetFloatingIPPoolID(d *Driver) (string, error)
 	GetInstancePortID(d *Driver) (string, error)
+	VolumeCreate(d *Driver) (string, error)
+	WaitForVolumeStatus(d *Driver, status string) error
+	VolumeAttach(d *Driver) (string, error)
 }
 
 type GenericClient struct {
-	Provider *gophercloud.ProviderClient
-	Compute  *gophercloud.ServiceClient
-	Network  *gophercloud.ServiceClient
+	Provider     *gophercloud.ProviderClient
+	Compute      *gophercloud.ServiceClient
+	Network      *gophercloud.ServiceClient
+	BlockStorage *gophercloud.ServiceClient
 }
 
 func (c *GenericClient) CreateInstance(d *Driver) (string, error) {
-	serverOpts := servers.CreateOpts{
+
+	var serverOpts servers.CreateOptsBuilder
+
+	var networks []servers.Network
+	if d.NetworkId != "" {
+		networks = []servers.Network{
+			{
+				UUID: d.NetworkId,
+			},
+		}
+	}
+
+	serverOpts = &servers.CreateOpts{
 		Name:             d.MachineName,
 		FlavorRef:        d.FlavorId,
 		ImageRef:         d.ImageId,
 		UserData:         d.UserData,
+		Networks:         networks,
 		SecurityGroups:   d.SecurityGroups,
 		AvailabilityZone: d.AvailabilityZone,
 		ConfigDrive:      &d.ConfigDrive,
 	}
-	if d.NetworkId != "" {
-		serverOpts.Networks = []servers.Network{
-			{
-				UUID: d.NetworkId,
-			},
-		}
+
+	serverOpts = &keypairs.CreateOptsExt{
+		CreateOptsBuilder: serverOpts,
+		KeyName:           d.KeyPairName,
 	}
 
 	log.Info("Creating machine...")
 
-	server, err := servers.Create(c.Compute, keypairs.CreateOptsExt{
-		CreateOptsBuilder: serverOpts,
-		KeyName:           d.KeyPairName,
-	}).Extract()
+	var server *servers.Server
+	var err error
+
+	if d.BootFromVolume {
+		blockDevices := []bootfromvolume.BlockDevice{
+			bootfromvolume.BlockDevice{
+				BootIndex:           0,
+				DeleteOnTermination: true,
+				DestinationType:     bootfromvolume.DestinationVolume,
+				SourceType:          bootfromvolume.SourceImage,
+				UUID:                d.ImageId,
+				VolumeType:          d.VolumeType,
+				VolumeSize:          d.VolumeSize,
+			},
+		}
+		serverOpts = &bootfromvolume.CreateOptsExt{
+			CreateOptsBuilder: serverOpts,
+			BlockDevice:       blockDevices,
+		}
+		server, err = bootfromvolume.Create(c.Compute, serverOpts).Extract()
```

Comment:
```suggestion
		}
		if d.VolumeType != "" {
			c.Compute.Microversion = "2.67"
		}
		server, err = bootfromvolume.Create(c.Compute, serverOpts).Extract()
```

Reply:
Seems to be the correct approach to force the server to return the required API version 👍

Reply:
I like it, makes sense to me.